### DISCUSSION TITLE
Image extraction problem

### DISCUSSION THREAD

**Discussion by maivan-hoa** [pymupdf/PyMuPDF#4112] [Q&A]
I use the code below to extract images from pdf files. My pdf file has two pages, page 1 contains images and page 2 does not contain images. But when executing the code, use the `get_images` function to extract the image. Page 2 is also considered to have an image and that image if saved is the image of page 1. Has anyone ever encountered this problem before? I have attached the code and pdf file below, hoping to receive everyone's help.

- Code:
```python
def check_picture(file_path):
    with fitz.open(file_path) as doc:
        for page_num in range(len(doc)):
            contain_image = False
            page = doc[page_num]

            image_list = page.get_images(full=True)
            number_image = 0
            for img_index, img in enumerate(image_list):
                xref = img[0]
                img_info = doc.extract_image(xref)
                # print(img_info)
                number_image += 1
                contain_image = True
            print(f'Page {page_num+1}, contain image: {contain_image}')
            print(f'Page {page_num+1}, number image: {number_image}')
            print('-----')

check_picture('Picture_problem.pdf')
```

- Result
```
Page 1, contain image: True
Page 1, number image: 1
-----
Page 2, contain image: True
Page 2, number image: 1
-----
```
- File:
[Picture_problem.pdf](https://github.com/user-attachments/files/18003854/Picture_problem.pdf)

**Comment by JorjMcKie** [comment]
There is no real problem here!
When you read the documentation you will find that methods `page.get_images()` and `page.get_fonts()` **_only_** show the existing references to images / fonts in the page definition.
They do **_not_** state that these resources are actually **_used_** when the page is displayed.
To see which images are actually displayed, use `page.get_image_info()` -- or `page.get_text("dict")` and look at block type 1.

**Reply by maivan-hoa** [reply]
`the existing references to images / fonts in the page definition.` What does it mean? Can you explain more?

**Reply by JorjMcKie** [reply]
The page definition as a PDF (!) object can be printed using `print(doc.xref_object(page.xref))`. Inside there you will see `/Resources`. Underneath this you will see the resources available to (not necessarily **_used by_**) the page:
```python
 doc=pymupdf.open("Picture_problem.pdf")
page=doc[1]
print(doc.xref_object(page.xref))
<<
  /Type /Page
  /Parent 8 0 R
  /Resources 15 0 R
  /MediaBox [ 0 0 612 792 ]
  /Group <<
    /S /Transparency
    /CS /DeviceRGB
    /I true
  >>
  /Contents 6 0 R
>>
print(doc.xref_object(15))  # content of /Resources above
<<
  /Font 14 0 R
  /XObject <<
    /Im4 4 0 R   # an image - which simple is not displayed by the page
  >>
  /ProcSet [ /PDF /Text /ImageC /ImageI /ImageB ]
>>
```

**Reply by maivan-hoa** [reply]
Thank you for your enthusiastic support. I want to ask one more question. Hope to receive help from someone with experience like you. Thank you very much.
I have a pdf file. When extracting text, the resulting text has font errors and the text cannot be obtained as displayed. It seems that the text in the pdf file is displayed with embedded fonts. Is there any way I can extract text from such pdf files effectively?

- Here is the code that I use:
```python
import fitz

def extract_text_from_pdf(pdf_path):
    text = ""

    with fitz.open(pdf_path) as pdf_document:
        for page_num in range(pdf_document.page_count):
            page = pdf_document[page_num]
            text += page.get_text()
    
    return text

pdf_path = "Page_test.pdf"
text = extract_text_from_pdf(pdf_path)
print(text)
```

- This is the pdf file from which I want to extract data:
[Page_test.pdf](https://github.com/user-attachments/files/18021599/Page_test.pdf)

- Text in file:
![image](https://github.com/user-attachments/assets/e8c6058d-f48b-405e-b83e-f8c0401e38ed)

- Extracted text:
```
VI 
v¨n kiÖn ®¶ng toµn tËp  
 
 
 
Lêi giíi thiÖu tËp 2 
 V¨n kiÖn §¶ng Toµn tËp, TËp 2 n¨m 1930, ph¶n ¸nh ùa 
ra ®êi vµ ho¹t ®éng cña §¶ng trong phong trµo c¸c m¹ng ®Çu 
tiªn cña lÞch sö ViÖt Nam hiÖn ®¹i. 
C¸c v¨n kiÖn ë tËp nµy gåm ba chñ ®Ò lín: 
Mét lµ, vÒ sù ra ®êi cña §¶ng Céng s¶n ViÖt Nam, bao 
gåm tªn gäi, c−¬ng lÜnh, ®−êng lèi chiÕn l−îc, s¸ch l−îc vµ 
§iÒu lÖ cña §¶ng. 
Hai lµ, vÒ tæ chøc vµ ph−¬ng h−íng ho¹t ®éng cña c¸c 
®oµn thÓ chÝnh trÞ vµ MÆt trËn d©n téc thèng nhÊt do §¶ng 
ta l·nh ®¹o. 
Ba lµ, vÒ sù l·nh ®¹o vµ chØ ®¹o cña §¶ng ®èi víi cao 
trµo c«ng - n«ng chèng ®Õ quèc, phong kiÕn, t− s¶n ®ang diÔn 
ra réng kh¾p toµn quèc, nhÊt lµ ë NghÖ An vµ Hµ TÜnh. 
Trong tËp V¨n kiÖn nµy cßn cã mét sè tµi liÖu ph¶n ¸nh 
sù l·nh ®¹o cña ®ång chÝ NguyÔn ¸i Quèc ®èi víi §¶ng Céng 
s¶n ViÖt Nam vµ phong trµo c¸ch m¹ng ViÖt Nam, vÒ sù ñng 
hé cña §¶ng Céng s¶n ViÖt Nam ®èi víi Liªn X« - nhµ n−íc 
c«ng n«ng ®Çu tiªn trªn thÕ giíi. 
V¨n kiÖn §¶ng Toµn tËp, TËp 2 gi÷ mét vÞ trÝ quan träng 
®Æc biÖt ë chç nã ®¸nh dÊu mét giai ®o¹n míi, cã tÝnh chÊt 
b−íc ngoÆt c¨n b¶n cña c¸ch m¹ng v« s¶n thÕ giíi, ®ång thêi 
ph¶n ¸nh tinh thÇn ®éc lËp, tù chñ, c¸ch m¹ng vµ s¸ng t¹o 
cña §¶ng ta trong viÖc x¸c ®Þnh c−¬ng lÜnh, ®−êng lèi, chÝnh 
s¸ch vµ tæ chøc s¾p xÕp lùc l−îng c¸ch m¹ng ë ViÖt Nam 
ngay tõ khi §¶ng míi thµnh lËp. 
TËp V¨n kiÖn nµy cã nhiÒu ch÷ viÕt t¾t, nhiÒu b¶n mê 
g©y rÊt nhiÒu khã kh¨n trong qu¸ tr×nh ®èi chiÕu x¸c minh. 
Chóng t«i ®· cè g¾ng b¶o ®¶m ®é chÝnh x¸c cña v¨n b¶n, 
song khã cã thÓ tr¸nh khái nh÷ng thiÕu sãt, mong nhËn ®−îc 
sù gãp ý phª b×nh cña b¹n ®äc. 
 
Th¸ng 12 n¨m 1998 
Nhµ xuÊt b¶n chÝnh trÞ quèc gia 
V 
```

**Reply by JorjMcKie** [reply]
The font(s) in this file have invalid information for back-translating the visual information (glyphs) back to Unicode.
I tried multiple extractors with the same result.
Your only potential resort is trying OCR.

**Reply by maivan-hoa** [reply]
I have a pdf file, although there are no tables in the file, but when using the `find_tables()` function, the results can still extract the table. How should I go about determining exactly whether a pdf page contains a table or not? Looking forward to receiving your help.

- Here is the code that I use:
```python
def check_pdf_contain_table(file_path):
    with fitz.open(file_path) as doc:
        for page_num in range(len(doc)):
            contain_table = False
            page = doc[page_num]

            tabs = page.find_tables()
            if tabs.tables:
                contain_table = True
            print(f'Page {page_num+1}, contain table: {contain_table}')
            print('-----')
```

- This is the pdf file:
[Toán học.pdf](https://github.com/user-attachments/files/18148341/Toan.h.c.pdf)

**Reply by JorjMcKie** [reply]
Use `page.find_tables(strategy="lines_strict")`. This avoids using background color rectangles as table cell delimiters. Only real lines will be accepted.
